renku_notebooks/util/clean_user_registry_secrets.py

```diff
@@ -0,0 +1,164 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Scripts used to remove user registry secrets in k8s"""
+
+from datetime import datetime, timedelta
+from pathlib import Path
+import argparse
+import re
+import logging
+
+from kubernetes import client
+from kubernetes.config.incluster_config import (
+    SERVICE_CERT_FILENAME,
+    SERVICE_TOKEN_FILENAME,
+    InClusterConfigLoader,
+)
+import escapism
+
+
+def find_pod_by_servername(namespace, servername, k8s_client):
+    """Find the podname based on the servername annotation used by jupyterhub"""
+    try:
+        pod_list = k8s_client.list_namespaced_pod(namespace)
+    except client.rest.ApiException:
+        logging.warning(
+            f"Namespace {namespace} does not exist or the cluster is unreachable."
+        )
+    else:
+        for pod in pod_list.items:
```

Comment:
@ableuler I decided to name the registry secrets using a `uuid4` from python and then appending `-registry` at the end. This way I have no way of accidentally overwriting a secret and I can always find the matching user JH pod based on the labels as you suggested. Let me know what you think.

Reply:
If you agree with this approach and all the other changes all I have to do is edit the tests to make sure they work with the new changes. This is an easy change though. But if I need to do other changes I would rather tackle those before finalizing the tests.

Reply:
I think having the username still in the secret name is not bad in case you're looking for something manually. So maybe `{escaped_username}-{registry}-{random-uuid}?`

Reply:
@ableuler I fixed the secret names as you suggested. On the other related/side issues the status is as follows:
- The ui does indeed use some of the annotations we wanted to move as labels. Lorenzo confirmed this (see https://swiss-data-science.slack.com/archives/G67N59QL8/p1603372964031300). What Lorenzo suggested is that we can remap the k8s labels back into annotations when we return data from the API so this would not require changes to UI code at all and we would not have to have duplicates.
- The labels have the same key prefixes as the annotations
- I keep using the `safe_username` version of the username all throughout. I.e. both in labels and annotations, as well as in the pod and secret names (here we do not have a choice and we have to do this).

Let me know if you want me to do the label/annotation remapping as I mentioned in point 1. Otherwise this issue may be best tackled as a subtask of the larger task to implement schemas for renku-notebooks.

Reply:
If changes to the API require minor changes to UI (e.g. to keep the naming consistent and avoid confusion) it's fine.
Tbf I don't like the current object `annotation: { renku.io/<resource-name>: <value> }` structure, because the naming reflects the source name and any change makes it weird, like in this case where returning label data in an annotation object sounds confusing.

We can use generic names, like `repository` instead of `annotation`, and remove the `renku.io`  prefix. The `hub.jupyter.org/servername` could be moved one level up as `servername`.
```
repository: {
    namespace: <value>,
    commit: <value>,
    ...
},
servername: <value>
```

Reply:
I think the question here is to what extent we want to just forward the K8s object, and how much we want to have "our own" resource that potentially deviates from the k8s resource. Following @lorenzo-cavazzi s suggestion would allow us to modify properties on the k8s object without changing the notebook-service API. On the other hand it's super obvious now where an annotation comes from which can make debugging errors maybe a bit easier? Either way, I think #421 is the proper place to discuss this.